enhancements/multi-arch/heterogeneous-architecture.md

```diff
@@ -0,0 +1,348 @@
+podAffinity rule with nodeselectorterms. The nodeselectorterms will be defaulted with the architectures that the operator supports, but would
+also have a provision for users to specify the architectures they want (from the set of supported architectures).
+
+### Install
+
+For Phase 1, the way to add more workers to the cluster is to handcraft a machineset manifest file for a worker node and specify the architecture
+specific details like instance type, image references and deploy the machineset to get a worker node. This can be pseudo automated further by creating
+the manifests using `openshift-install create manifests` and modifying or adding worker manifest files. Eventually the installer should support being
+able to declare differing compute and control plane sections in the install-config mainly allowing architectures to be different. In doing so it would
+also generate appropriate machineset and machineconfig files tailored to the specific architecture. The machineset file would contain the instance
+types and region corresponding to the architecture chosen for the workers.
+
+### Open Questions
+
+- What would be the method to provide recommendation updates when moving from a homogeneous graph to a heterogeneous graph ?
+
+### Test Plan
+
+Testing for Phase 1:
+- QE would need to test heterogeneous nightly payloads before the entry is added to Cincinnati. This implies that the heterogeneous payloads will be
+released a bit later than the homogeneous payloads.
+- QE would test heterogeneous payloads on Azure and AWS platforms with the cluster configuration mainly being x86 control plane and ARM worker nodes.
+- Testing scenarios would include machine set scaling, upgrades involving ARM nodes and OVN testing on heterogeneous clusters.
+- New release payload acceptance jobs on heterogeneous clusters.
+
+Testing for Phase 2:
+- With the addition of Power and Z architectures, there will be more permutations to test. We would need to figure out the absolute essential set of
+permutations and the tests to run on those.
+- Imagestream conformance tests.
+- OLM and console testing
+- Test various upgrade scenarios (homogeneous -> heterogeneous and vice versa)
+
+### Graduation Criteria
+
+Phase 1 will have a pre-4.11 dev preview and 4.11 tech preview with documented limitations.
+
+#### Dev Preview -> Tech Preview
+
+There will be a tech preview of basic functionality (Phase 1 as described above) in 4.11
+
+#### Tech Preview -> GA
+
+TBD
+
+#### Removing a deprecated feature
+
+This enhancement does not describe removing a feature of Openshift, this section is not applicable.
+
+### Upgrade / Downgrade Strategy
+
+The upgrade strategy is described in the design details section above. The important changes here are having a separate architecture
+for the heterogeneous releases and also the work needed to support upgrades from a homogeneous release -> heterogeneous release and
+vice versa in the future.
+
+### Version Skew Strategy
+
+TBD
```

Comment:
Also somewhat in the version-skew front would be "updates" between homogeneous and heterogeneous releases.  Obviously if a cluster had attached heterogeneous compute, we'd want to try and block heterogeneous-to-homogeneous transitions to avoid breaking that compute.  But that's always going to be at least a bit racy, e.g. if someone launches their first alt-arch compute node after initiating the transition to a homogeneous release image but before the CVO had gotten deep enough in to move the machine-(config-)API portions over to homogeneous images.  What would happen to those machines?  Would the lock when the MCO tries to update them into a homogeneous (and incompatible) RHCOS?  And then maybe a MachineHealthCheck would trigger the machine-API components to gracefully reap the alt-arch nodes?  Probably not too many admins would attempt something like this, but it might be worth at least talking through the broad strokes of how we expect this to work out.

Reply:
the upgrade of machine-os-content on that node will fail, but i'm not sure if machine-api will reap the node. The other thing here is, if the alt-arch node comes up as schedulable (while on the previous heterogeneous release), some operators (for the homogeneous release) could be scheduled on that node causing them to fail and thus failing the upgrade in that state even before getting to the above mentioned scenario.

Reply:
mentioned this in version skew although not sure what can be dine here other than the admin scaling down that machineset manually and restarting the upgrade. slightly related - the machine api folks are also having a conversation on whether machinesets' instance type and image id should be allowed to be modified post installation.